### PR TITLE
Migrate SOAP service to REST service

### PR DESCRIPTION
**Pull request by haocheng**
Hi,

I have migrated JIRA SOAP service to REST service and I have manually tested that all existing functions work as expected. Please let me know if there are any problems.

Cheers

### CODE REVIEW

File: src/main/java/hudson/plugins/jira/JiraRestService.java [hunk 1]
```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2012 MeetMe, Inc.
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/jira/JiraRestService.java]: 2012?

File: src/main/java/hudson/plugins/jira/versionparameter/JiraVersionParameterDefinition.java [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.regex.Pattern;
 
 import static hudson.Util.fixNull;
-import static java.util.Arrays.asList;
 
 public class JiraVersionParameterDefinition extends ParameterDefinition {
     private static final long serialVersionUID = 3927562542249244416L;
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/jira/versionparameter/JiraVersionParameterDefinition.java]: Update serial?

File: src/main/java/hudson/plugins/jira/JiraRestService.java (hunk 1 above)
Comment (haocheng) [inline src/main/java/hudson/plugins/jira/JiraRestService.java]: Sorry for the typo. Fixed with correct year and owner.

Thank you!

File: src/main/java/hudson/plugins/jira/versionparameter/JiraVersionParameterDefinition.java (hunk 2 above)
Comment (haocheng) [inline src/main/java/hudson/plugins/jira/versionparameter/JiraVersionParameterDefinition.java]: Fixed. Thanks!

**Comment by mitchross** [comment]
Thanks for this!

**Comment by gmale** [comment]
+1, our project is very close to being impacted by the upcoming [SOAP deprecation](https://developer.atlassian.com/jiradev/jira-apis/jira-api-status/jira-rpc-services/jira-xml-rpc-overview/soap-and-xml-rpc-api-deprecation-notice). This PR is right on time! :+1:

**Comment by scott-compsource** [comment]
Hi, we also use Jira and depend upon the Jenkins plugin. Thank you very much! 

I am new to Jenkins plugin development, so how do I get the plugin installed into Jenkins? 

On the Jenkins plugin page the options to 1) Upload Plugin: "You can upload a .hpi file to install a plugin from outside the central plugin repository." but I don't see a .hpi or .jpi in the Git repository. 

2) Update Site: Where you specify a URL. The default from Jenkins is: http://updates.jenkins-ci.org/update-center.json. What value to use to uploading this "Migrate SOAP service to REST service#58" ?

**Comment by haocheng** [comment]
Hi @scott-compsource ,

If you can wait for a while, the best way is to wait for the PR merged into master. Then you can just upgrade the plugin as usual. 

If you need to upgrade now, you can git clone migrate-soap-to-rest branch and run: 
```
mvn clean install
```

If maven completes without error message, you can find jira.hpi under target/ folder. You can upload the file to Jenkins using 1)  method you mentioned.

Hope this helps!

Cheers

**Comment by scott-compsource** [comment]
Haocheng, Maven built successfully and deployed plugin to Jenkins and it worked great. Thank you very much!!!

**Comment by haocheng** [comment]
@scott-compsource 

Glad it helps! Please let me know if you find any issues.

Cheers

**Comment by harcher81** [comment]
Thank for this, but I tried this pull request, but I have a problem with it. After 1h, on Windows, all sockets was used by java.exe and jenkins have 22000 threads. I have rollback to the official version and it works.

**Comment by haocheng** [comment]
Hi @harcher81 ,

I will try to check what might be wrong, but the Jenkins instance seems to be fine on my Ubuntu server. Can you provide more information about the issue, such as the thread dump? 
https://wiki.jenkins-ci.org/display/JENKINS/Obtaining+a+thread+dump

Thank you!

**Comment by yaytul** [comment]
Is there an ETA for merging this PR? We have couple projects affected by this.

**Comment by warden** [comment]
This is tracked here: https://issues.jenkins-ci.org/browse/JENKINS-18166
This is a major change and needs to be tested thoroughly and as it seems, the first issue was discovered. SOAP deprecation will not happen until JIRA 7 which is not soon.

**Comment by olamy** [comment]
why not keeping both connectivity modes and made a configuration to switch from SOAP to REST?
Note: Soap must be default

**Comment by warden** [comment]
That's actually a good idea to keep backwards compatibility!
@haocheng do you think you could refactor this to keep  SOAP and set this via an option with preselected SOAP?

**Comment by haocheng** [comment]
Hi @olamy @warden ,

Sorry for late reply due to timezone difference.

I think I can try, but I am not sure if it's worth the complexity to maintain both versions since the SOAP API will be removed in the near future (and it should have been removed in JIRA Cloud though it seems still worked last time I checked). If we keep both SOAP/REST API, then we need to implement the same function twice every time we want to add a new feature... Why not just keep the existing 1.x branch for people who want to stay with SOAP API? We can create 2.x branch with REST API only. 

Cheers

**Comment by haocheng** [comment]
Hi @harcher81 ,

I have restarted my Jenkins and monitored it for more than 24 hours, but I did not see the thread issue you mentioned. Can you provide more details such as your Jenkins version, Windows OS version? It would be easier for me to debug.

Thank you!

**Comment by warden** [comment]
SOAP will be deprecated eventually but many people are using older JIRA versions and are unable to upgrade for different reasons. Having branches would mean a bigger maintain costs to introduce fixes and features so this is not the way to go.

**Comment by harcher81** [comment]
@haocheng 
I can reproduce it on my workstation (Windows 7), Windows 8 and on Windows Server 2012R2. Nothing special to see in the threadDump, but on the ressource monitor I see an anormal numbers of threads.

With the command `netstat -a -o -n` there are a huge numbers of sockets associated to the Jenkins process.

To reproduce it, I use a Jenkins with this version of the JIRA plugin. I configure the JIRA sites. I create a job with the "Update relevant JIRA issues" and for each build I have a socket used by the plugin.

**Comment by haocheng** [comment]
Hi @harcher81 ,

Thank you for providing more details! I will try to install a Windows VM and test this issue. Anyone saw the same issue on Linux/Mac? 

Cheers

**Comment by mulmschneider** [comment]
@warden just FYI, SOAP won't "be deprecated eventually" - it's already disabled for Jira Cloud.
(We had to disable the jira-plugin on our buildserver because of this.)

**Comment by haocheng** [comment]
@mulmschneider 

Sorry I may need some more time for finding the bug on Windows machine. If you need to use the JIRA plugin with REST API now, you can build the plugin by clone this branch. I have been using it for a few weeks and it seems working fine on my Ubuntu server. 

Cheers

**Comment by harcher81** [comment]
I think it depends of this bug: https://issues.apache.org/jira/browse/HTTPCLIENT-1668 .
On https://github.com/jenkinsci/jira-plugin/pull/58/files#diff-11ef1cf693a1a3413bcf3fcf87b8904cR303(line 303), there is no realy connection timeout.

**Comment by warden** [comment]
@haocheng maybe you can refactor to preserve the old functionality for now? this way we could merge it quicker and for the people having problems, they can use the old SOAP version

**Comment by haocheng** [comment]
@harcher81 

Interesting... Thank you for the information. I will try to downgrade the httpcomponents library and see if that fixed the issue.

Cheers

**Comment by haocheng** [comment]
@warden 

I will try my best but that seems to be more difficult than I expected. For people using JIRA cloud, keep SOAP

**Comment by gmale** [comment]
Looks like the SOAP API was officially removed from JIRA cloud today. Our builds are broken and we're writing some scripts to work around it . . . in hopes that this plugin will be updated soon.

**Comment by the4yeast** [comment]
Same issue for us here, hoping the plugin can support the REST API soon.

**Comment by warden** [comment]
@haocheng seems that create release version does not work,seems like a dependency error.
Looking at dependency:tree there are multiple httpclient versions referenced.
Can you take a look?

```java
ERROR: Publisher 'Create JIRA version' aborted due to exception: 
java.lang.NoSuchFieldError: INSTANCE
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.<clinit>(SSLConnectionSocketFactory.java:144)
	at org.apache.http.client.fluent.Executor.<clinit>(Executor.java:78)
	at org.apache.http.client.fluent.Request.execute(Request.java:151)
	at hudson.plugins.jira.JiraRestService.getVersions(JiraRestService.java:148)
	at hudson.plugins.jira.JiraSession.getVersions(JiraSession.java:99)
	at hudson.plugins.jira.JiraSite.getVersions(JiraSite.java:439)
	at hudson.plugins.jira.JiraVersionCreator.perform(JiraVersionCreator.java:77)
	at hudson.tasks.BuildStepMonitor$3.perform(BuildStepMonitor.java:45)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:776)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:670)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:381)
```

**Comment by joelfire** [comment]
FYI I have a fix and created a pull request for https://github.com/haocheng/jira-plugin.

 

The http 4.5 components conflicted with the 4.2 components in maven-plugin. Removing 2 explicit dependencies and backing fluent-hc to 4.2.1 fixed it.

 

Cheers,

 

-j

**Comment by warden** [comment]
Great job guys, this has been merged into master. If you can build and install from master branch and let me know if there are any issues, that'd be great. 
I hope we'll release v. 2.0 early September.

**Comment by haocheng** [comment]
@warden @joelfire 

Cool! Thank you for your help!

P.S. Sorry I don't have enough time to fix this issue while keeping both SOAP/REST API...

**Pull request merged**